### PR TITLE
Ingest lnet data from storage servers into management db

### PR DESCRIPTION
**Pull request by johnsonw**
In addition to collecting network interface information we should also
collect lnet data from `lnetctl net show` from each storage server. This
information should then be ingested into the database on the manager
through the iml-network service.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

---
This change is [Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/2362)

### CODE REVIEW

**Comment by johnsonw** [comment]
Sample output from the postgres database after ingesting lnet data from the storage servers:

```
chroma=> select * from lnet;
 id  |       nid       | host_id | net_type | status | interfaces
-----+-----------------+---------+----------+--------+------------
 751 | 10.73.20.22@tcp |       4 | tcp      | up     | {eth1}
 752 | 10.73.20.21@tcp |       3 | tcp      | up     | {eth1}
 753 | 10.73.20.11@tcp |       1 | tcp      | up     | {eth1}
 754 | 10.73.20.12@tcp |       2 | tcp      | up     | {eth1}
```

File: iml-agent/src/network_interfaces.rs [hunk 1]
```diff
@@ -38,3 +46,51 @@ pub async fn get_interfaces() -> Result<Vec<NetworkInterface>, ImlAgentError> {
...
+    fn test_parse_lnetctl_net_show_output() {
+        let data = r#"net:
+    - net type: lo
+      local NI(s):
+        - nid: 0@lo
+          status: up
+    - net type: tcp
+      local NI(s):
+        - nid: 10.73.20.21@tcp
+          status: up
```
Comment (ip1981) [inline iml-agent/src/network_interfaces.rs:74]: nit: could have one or two with `status: down`

Comment (johnsonw) [reply iml-agent/src/network_interfaces.rs:74]: Sounds good, i'll add that.

**Review by ip1981 — DISMISSED** [review]

**Comment by johnsonw** [comment]
Two tables are created and updated according to the lnet data spanning across the storage servers:
1. lnet
2. nid

The nid table contains all of the nid information for a particular host. There will likely be multiple entries for a single host. The lnet table on the other hand, will have a single entry for each host and contain a list of nid id's the correspond to the matching nids for the host in context. 

Example output:

```
chroma=> select * from lnet;
 id | host_id | nids
----+---------+------
  3 |       6 | {6}
  4 |       5 | {1}
  1 |       7 | {7}
  2 |       8 | {8}
(4 rows)

chroma=> select * from nid;
 id | net_type | host_id |       nid       | status | interfaces
----+----------+---------+-----------------+--------+------------
  6 | tcp      |       6 | 10.73.20.12@tcp | up     | {eth1}
  1 | tcp      |       5 | 10.73.20.11@tcp | up     | {eth1}
  7 | tcp      |       7 | 10.73.20.21@tcp | up     | {eth1}
  8 | tcp      |       8 | 10.73.20.22@tcp | up     | {eth1}
(4 rows)
```

File: iml-services/iml-network/src/main.rs [hunk 2]
```diff
@@ -95,56 +67,277 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+
+    // Update the lnet table with the list of nids for the current host_id
+    sqlx::query!(
+        r#"
+            INSERT INTO lnet
+            (host_id, nids)
+            VALUES($1, $2)
+            ON CONFLICT (host_id)
+                DO
+                UPDATE SET nids = EXCLUDED.nids"#,
```
Comment (ip1981) [inline iml-services/iml-network/src/main.rs:215]: Wouldn't it be better within a transaction (two or more queries)?

**Comment by johnsonw** [comment]
As discussed, I added `state` to the lnet table. Here is the new output:
```
chroma=> select * from lnet;
 id | host_id | state | nids
----+---------+-------+------
  4 |       4 | up    | {4}
  1 |       2 | up    | {1}
  2 |       1 | up    | {2}
  3 |       3 | up    | {3}
(4 rows)

chroma=> select * from nid;
 id | net_type | host_id |       nid       | status | interfaces
----+----------+---------+-----------------+--------+------------
  4 | tcp      |       4 | 10.73.20.22@tcp | up     | {eth1}
  1 | tcp      |       2 | 10.73.20.12@tcp | up     | {eth1}
  2 | tcp      |       1 | 10.73.20.11@tcp | up     | {eth1}
  3 | tcp      |       3 | 10.73.20.21@tcp | up     | {eth1}
(4 rows)
```

File: chroma_core/models/host.py [hunk 3]
```diff
@@ -1163,6 +1165,22 @@ def run(self, kwargs):
...
+                [host.id],
+            )
+            cursor.execute(
+                """
+                DELETE FROM nid
+                WHERE host_id = %s
+                """,
+                [host.id],
+            )
+
```
Comment (jgrund) [inline chroma_core/models/host.py:1182]: Should also delete from network here.

File: migrations/20201102184120_lnet.sql [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    id                  serial PRIMARY KEY,
+    host_id             int NOT NULL UNIQUE,
+    state               text NOT NULL,
+    nids                int[] NOT NULL
+);
+
+CREATE TABLE IF NOT EXISTS nid (
+    id serial PRIMARY KEY,
+    net_type text NOT NULL,
+    host_id int NOT NULL UNIQUE,
```
Comment (jgrund) [inline migrations/20201102184120_lnet.sql:11]: Should this really be unique?

Comment (johnsonw) [reply migrations/20201102184120_lnet.sql:11]: I thought when we discussed earlier we said it should be unique. Any reason why it wouldn't be?

Comment (jgrund) [reply migrations/20201102184120_lnet.sql:11]: because then you could only have one host_id regardless of how many nids were on that host.

**Review by jgrund — APPROVED** [review]

**Review by nlinker — APPROVED** [review]

**Pull request merged**